README.md
# resource_manager

This is a library of Flutter for resource manager works as similar as what Android does. It works with JSON file.  

使用JSON文件的类Android的Flutter资源管理库。

![alt](https://github.com/chrisfang6/uploads/blob/master/resource_manager/1.png)

## How to use

1. Use `initBaseResources(double devicePixelRatio, String baseResourceFile)` to init the base resources from JSON file.  

使用`initBaseResources(double devicePixelRatio, String baseResourceFile)`从指定JSON文件初始化基础资源。  

eg.,`initBaseResources(window.devicePixelRatio, 'res/values.json')`  

2. Use `getResources(String jsonFile)` to get resources from JSON file.   

使用`getResources(String jsonFile)`从JSON文件获取资源。  

eg., `getResources('res/example.json')`  

3. Use `getXXX(String name, {Resources fromResources, Set<String> soughtNames})` to get the resource by name.   

使用`getXXX(String name, {Resources fromResources, Set<String> soughtNames})`获取指定名称资源。

### JSON file

The JSON file for resources has nodes named **dimens**, **colors**, **drawable**, **strings**, **integers**, **doubles**.   

资源JSON文件的节点有 **dimens**, **colors**, **drawable**, **strings**, **integers**, **doubles**。   

eg., *value.json*

```json
{
    "dimens": {
        "h1_text_size": "32dp",
        "h2_text_size": "24dp",
        "h3_text_size": "19dp",
        "h4_text_size": "16dp",
        "h5_text_size": "14dp",
        "h6_text_size": "13dp"
    },
    "colors": {
        "black": "#000000",
        "FireBrick": "#B22222",
        "DeepPink": "#FF1493",
        "Gold": "#FFD700",
        "OliveDrab": "#6B8E23",
        "Lime": "#00FF00",
        "RoyalBlue": "#4169E1",
        "DarkOrchid": "#9932CC"
    },
    "drawable": {
        "logo": "graphics/FlutterDart.png"
    },
    "strings": {
        "string 1": "base string 1"
    },
    "integers": {
        "integer 1": 10,
        "integer 2": 20
    },
    "doubles": {
        "double 1": 30.3,
        "double 2": 40
    }
}
```



And *example.json*

```json
{
    "dimens": {
        "example_text_size": "@dimen/to base h2",
        "to base h2": "@dimen/h2_text_size",
        "color_bar_height": "30dp",
        "color_bar_padding": "5dp",
        "list_padding": "20dp",
        "gap": "50dp",
        "shown dp": "50dp",
        "shown px": "50px"
    },
    "colors": {
        "As FireBrick": "@color/to base fire brick",
        "to base fire brick": "@color/FireBrick",
        "H1 color": "@color/RoyalBlue",
        "H2 color": "@color/RoyalBlue",
        "H3 color": "@color/RoyalBlue",
        "H4 color": "@color/RoyalBlue",
        "H5 color": "@color/RoyalBlue",
        "H6 color": "@color/RoyalBlue",
        "image label color": "@color/DarkOrchid",
        "num label color": "@color/Gold",
        "pixel label color": "@color/Lime",
        "string label color": "@color/OliveDrab"
    },
    "drawable": {
        "logo": "graphics/flutter.PNG"
    },
    "strings": {
        "string 1": "example string 1"
    },
    "integers": {
        "integer 1": "@integer/integer 3",
        "integer 2": 21,
        "integer 3": "@integer/integer 1"
    },
    "doubles": {
        "double 1": "@double/double 1",
        "double 2": 41
    }
}
```



Value can be a string with prefix, such **@dimen/**, **@color/**, **@drawable/**, **@string/**, **@integer/** and **@double/**, and then refer to another resource. If the resource can't be find in the referred resources, it would keep looking for the resource in the base resources.  

值可以是有  **@dimen/**, **@color/**, **@drawable/**, **@string/**, **@integer/** 和 **@double/** 为前缀的字符串，指向其他资源。当在指定资源文件中找不到指定名称资源时，继续在基础资源文件中寻找。

### Methods for getting resource

1. Dimen: `getDimension(String name, {Resources fromResources, Set<String> soughtNames})`

2. Color: `getColor(String name, {Resources fromResources, Set<String> soughtNames})`

3. Drawable: `getDrawable(String name, {Resources fromResources, Set<String> soughtNames})`

4. String: `getString(String name, {Resources fromResources, Set<String> soughtNames})`

5. Integer: `getInteger(String name, {Resources fromResources, Set<String> soughtNames})`

6. Double: `getDouble(String name, {Resources fromResources, Set<String> soughtNames})`

If optional parameter `fromResources` is ignored, then the methods would try to get resources from base resource file.   

如果忽略可选参数`fromResources`，则直接从基础资源文件中寻找。

Ignore optional parameter `soughtNames`, please.   

请忽略可选参数`soughtNames`。

CHANGELOG.md

## ![1.0.3](https://img.shields.io/badge/Release-1.0.3-success.svg)- ![clock](https://github.com/chrisfang6/uploads/blob/master/clock.svg) 04/03/2019.

* Change license to Apache License 2.0

## ![1.0.2](https://img.shields.io/badge/Release-1.0.2-success.svg)- ![clock](https://github.com/chrisfang6/uploads/blob/master/clock.svg) 16/04/2019.

* Change license to BSD 3-clause.

## ![1.0.1](https://img.shields.io/badge/Release-1.0.1-success.svg) - ![clock](https://github.com/chrisfang6/uploads/blob/master/clock.svg) 26/03/2019.

* Add type annotations

## ![1.0.0](https://img.shields.io/badge/Release-1.0.0-success.svg) - ![clock](https://github.com/chrisfang6/uploads/blob/master/clock.svg) 21/03/2019.

* Including 'dimens', 'colors', 'drawable', 'strings', 'integers' and 'doubles'
* With example app.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

// import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

// import 'package:example_resource_manager/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // // Build our app and trigger a frame.
    // await tester.pumpWidget(MyApp());

    // // Verify that our counter starts at 0.
    // expect(find.text('0'), findsOneWidget);
    // expect(find.text('1'), findsNothing);

    // // Tap the '+' icon and trigger a frame.
    // await tester.tap(find.byIcon(Icons.add));
    // await tester.pump();

    // // Verify that our counter has incremented.
    // expect(find.text('0'), findsNothing);
    // expect(find.text('1'), findsOneWidget);
  });
}


test/resource_manager_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:resource_manager/resource_manager.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  //   expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  // });
}


lib/resources.dart
/*
 * Copyright 2019 chrisfang6
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import 'package:json_annotation/json_annotation.dart';

part 'resources_g.dart';

@JsonSerializable(nullable: false)
class Resources {
  @JsonKey(name: 'dimens', nullable: true)
  Map<String, String> dimens;

  @JsonKey(name: 'colors', nullable: true)
  Map<String, String> colors;

  @JsonKey(name: 'drawable', nullable: true)
  Map<String, String> drawable;

  @JsonKey(name: 'strings', nullable: true)
  Map<String, String> strings;

  @JsonKey(name: 'integers', nullable: true)
  Map<String, dynamic> integers;

  @JsonKey(name: 'doubles', nullable: true)
  Map<String, dynamic> doubles;

  Resources(
      {this.dimens,
      this.colors,
      this.drawable,
      this.strings,
      this.integers,
      this.doubles});

  factory Resources.fromJson(Map<String, dynamic> json) =>
      _$ResourcesFromJson(json);

  Map<String, dynamic> toJson() => _$ResourcesToJson(this);
}


lib/resources_g.dart
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'resources.dart';

// **************************************************************************
// JsonSerializableGenerator
// **************************************************************************

Resources _$ResourcesFromJson(Map<String, dynamic> json) {
  return Resources(
      dimens: (json['dimens'] as Map<String, dynamic>)?.map(
        (k, dynamic e) => MapEntry(k, e as String),
      ),
      colors: (json['colors'] as Map<String, dynamic>)?.map(
        (k, dynamic e) => MapEntry(k, e as String),
      ),
      drawable: (json['drawable'] as Map<String, dynamic>)?.map(
        (k, dynamic e) => MapEntry(k, e as String),
      ),
      strings: (json['strings'] as Map<String, dynamic>)?.map(
        (k, dynamic e) => MapEntry(k, e as String),
      ),
      integers: json['integers'] as Map<String, dynamic>,
      doubles: json['doubles'] as Map<String, dynamic>);
}

Map<String, dynamic> _$ResourcesToJson(Resources instance) => <String, dynamic>{
      'dimens': instance.dimens,
      'colors': instance.colors,
      'drawable': instance.drawable,
      'strings': instance.strings,
      'integers': instance.integers,
      'doubles': instance.doubles
    };


example/lib/style.dart
/*
 * Copyright 2019 chrisfang6
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import 'package:flutter/cupertino.dart';
import 'package:resource_manager/resource_manager.dart';

FontWeight _bold = FontWeight.bold;
FontWeight _medium = FontWeight.w500;
FontWeight _light = FontWeight.w300;

TextStyle getH1RegularTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h1_text_size"),
      fontFamily: "sans-serif",
      color: color,
    );
TextStyle getH1MediumTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h1_text_size"),
      fontFamily: "sans-serif",
      fontWeight: _medium,
      color: color,
    );
TextStyle getH1BoldTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h1_text_size"),
      fontFamily: "sans-serif",
      fontWeight: _bold,
      color: color,
    );
TextStyle getH1LightTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h1_text_size"),
      fontFamily: "sans-serif",
      fontWeight: _light,
      color: color,
    );
TextStyle getH2RegularTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h2_text_size"),
      fontFamily: "sans-serif",
      color: color,
    );
TextStyle getH2MediumTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h2_text_size"),
      fontFamily: "sans-serif",
      fontWeight: _medium,
      color: color,
    );
TextStyle getH3RegularTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h3_text_size"),
      fontFamily: "sans-serif",
      color: color,
    );
TextStyle getH3BoldTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h3_text_size"),
      fontFamily: "sans-serif",
      fontWeight: _bold,
      color: color,
    );
TextStyle getH4RegularTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h4_text_size"),
      fontFamily: "sans-serif",
      color: color,
    );
TextStyle getH4MediumTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h4_text_size"),
      fontFamily: "sans-serif",
      fontWeight: _medium,
      color: color,
    );
TextStyle getH5RegularTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h5_text_size"),
      fontFamily: "sans-serif",
      color: color,
    );
TextStyle getH6RegularTextStyle(Color color) => TextStyle(
      fontSize: getDimension("h6_text_size"),
      fontFamily: "sans-serif",
      color: color,
    );


lib/resource_manager.dart
/*
 * Copyright 2019 chrisfang6
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
library resource_manager;

import 'dart:async';
import 'dart:convert';

import 'package:flutter/services.dart';
import 'package:resource_manager/resources.dart';

const _prefixDimen = "@dimen/";
const _prefixColor = "@color/";
const _prefixDrawable = "@drawable/";
const _prefixString = "@string/";
const _prefixInteger = "@integer/";
const _prefixDouble = "@double/";

double _devicePixelRatio;
Resources _resources;

Future initBaseResources(
    double devicePixelRatio, String baseResourceFile) async {
  setDevicePixelRatio(devicePixelRatio);
  _setResources(Resources.fromJson(
      json.decode(await rootBundle.loadString(baseResourceFile))
          as Map<String, dynamic>));
}

Future<Resources> getResources(String jsonFile) async => Resources.fromJson(
    json.decode(await rootBundle.loadString(jsonFile)) as Map<String, dynamic>);

void setDevicePixelRatio(double devicePixelRatio) {
  _devicePixelRatio = devicePixelRatio;
}

void _setResources(Resources resources) {
  _resources = resources;
}

// int _dpToPxWithDensity(num dp) =>
//     (dp * _devicePixelRatio + (dp >= 0 ? 0.5 : (-0.5))).round();

int _pxToDpWithDensity(num px) =>
    (px / _devicePixelRatio + (px >= 0 ? 0.5 : (-0.5))).round();

/// The logic pixel by named key. Same as getDimensionPixelOffset().
double getDimension(String name,
    {Resources fromResources, Set<String> soughtNames}) {
  String prefix = _prefixDimen;
  dynamic value =
      (fromResources?.dimens != null ? fromResources.dimens[name] : null) ??
          (_resources?.dimens != null ? _resources.dimens[name] : null);
  if (value is! String) {
    return 0.0;
  }
  if (RegExp(r"^-{0,1}\d+(dp|sp)$").hasMatch(value)) {
    // dp or sp
    return int.tryParse(RegExp(r"-{0,1}\d+").firstMatch(value).group(0))
            ?.toDouble() ??
        0.0;
  } else if (RegExp(r"^-{0,1}\d+px$").hasMatch(value)) {
    // px
    return _pxToDpWithDensity(
      int.tryParse(RegExp(r"-{0,1}\d+").firstMatch(value).group(0)) ?? 0,
    ).toDouble();
  } else if (RegExp(r"^" + "$prefix").hasMatch(value) &&
      fromResources != null &&
      fromResources != _resources) {
    String referName = value.substring(prefix.length);
    soughtNames ??= Set<String>();
    soughtNames.add(name);
    if (!soughtNames.contains(referName)) {
      soughtNames.add(referName);
      return getDimension(
        referName,
        fromResources: fromResources,
        soughtNames: soughtNames,
      );
    } else if (_notBaseResources(fromResources)) {
      // The name has been sought in current resource file,
      // If the current resource is not the base resource,
      // seek the base resource.
      soughtNames.clear();
      return getDimension(
        referName,
        soughtNames: soughtNames,
      );
    }
  }
  return 0.0;
}

/// The Color by named key
int getColor(String name, {Resources fromResources, Set<String> soughtNames}) {
  String prefix = _prefixColor;
  dynamic value =
      (fromResources?.colors != null ? fromResources.colors[name] : null) ??
          (_resources?.colors != null ? _resources.colors[name] : null);
  if (value is! String) {
    return 0xffffffff;
  }
  if (value.startsWith('#')) {
    return int.tryParse('0x' + _validColor(value)) ?? 0;
  } else if (value.startsWith(prefix)) {
    String referName = value.substring(prefix.length);
    soughtNames ??= Set<String>();
    soughtNames.add(name);
    if (!soughtNames.contains(referName)) {
      soughtNames.add(referName);
      return getColor(
        referName,
        fromResources: fromResources,
        soughtNames: soughtNames,
      );
    } else if (_notBaseResources(fromResources)) {
      // The name has been sought in current resource file,
      // If the current resource is not the base resource,
      // seek the base resource.
      soughtNames.clear();
      return getColor(
        referName,
        soughtNames: soughtNames,
      );
    }
  }
  return 0xffffffff;
}

String _validColor(color) {
  String value = color.substring(1);
  if (value.length > 8) {
    value = value.substring(0, 8);
  }
  if (value.length < 8) {
    int less = 8 - value.length;
    for (int i = 0; i < less; i++) {
      value = 'f' + value;
    }
  }
  return value;
}

/// The Drawable by named key
String getDrawable(String name,
    {Resources fromResources, Set<String> soughtNames}) {
  String prefix = _prefixDrawable;
  dynamic value =
      (fromResources?.drawable != null ? fromResources.drawable[name] : null) ??
          (_resources?.drawable != null ? _resources.drawable[name] : null);
  if (value is! String) {
    return '';
  }
  if (value.startsWith(prefix)) {
    String referName = value.substring(prefix.length);
    soughtNames ??= Set<String>();
    soughtNames.add(name);
    if (!soughtNames.contains(referName)) {
      soughtNames.add(referName);
      return getDrawable(
        referName,
        fromResources: fromResources,
        soughtNames: soughtNames,
      );
    } else if (_notBaseResources(fromResources)) {
      // The name has been sought in current resource file,
      // If the current resource is not the base resource,
      // seek the base resource.
      soughtNames.clear();
      return getDrawable(
        referName,
        soughtNames: soughtNames,
      );
    }
  }
  return value;
}

/// The String by named key
String getString(String name,
    {Resources fromResources, Set<String> soughtNames}) {
  String prefix = _prefixString;
  dynamic value =
      (fromResources?.strings != null ? fromResources.strings[name] : null) ??
          (_resources?.strings != null ? _resources.strings[name] : null);
  if (value is! String) {
    return '';
  }
  if (value.startsWith(prefix)) {
    String referName = value.substring(prefix.length);
    soughtNames ??= Set<String>();
    soughtNames.add(name);
    if (!soughtNames.contains(referName)) {
      soughtNames.add(referName);
      return getString(
        referName,
        fromResources: fromResources,
        soughtNames: soughtNames,
      );
    } else if (_notBaseResources(fromResources)) {
      // The name has been sought in current resource file,
      // If the current resource is not the base resource,
      // seek the base resource.
      soughtNames.clear();
      return getString(
        referName,
        soughtNames: soughtNames,
      );
    }
  }
  return value;
}

/// The Integer by named key
int getInteger(String name,
    {Resources fromResources, Set<String> soughtNames}) {
  String prefix = _prefixInteger;
  dynamic value =
      (fromResources?.integers != null ? fromResources.integers[name] : null) ??
          (_resources?.integers != null ? _resources.integers[name] : null);
  // Found Number
  if (value is num) {
    return value.toInt();
  }
  if (value is String && value.startsWith(prefix)) {
    // Refer to other resource
    String referName = value.substring(prefix.length);
    soughtNames ??= Set<String>();
    soughtNames.add(name);
    if (!soughtNames.contains(referName)) {
      soughtNames.add(referName);
      return getInteger(
        referName,
        fromResources: fromResources,
        soughtNames: soughtNames,
      );
    } else if (_notBaseResources(fromResources)) {
      // The name has been sought in current resource file,
      // If the current resource is not the base resource,
      // seek the base resource.
      soughtNames.clear();
      return getInteger(
        referName,
        soughtNames: soughtNames,
      );
    }
  }
  return 0;
}

/// The Double by named key
double getDouble(String name,
    {Resources fromResources, Set<String> soughtNames}) {
  String prefix = _prefixDouble;
  dynamic value =
      (fromResources?.doubles != null ? fromResources.doubles[name] : null) ??
          (_resources?.doubles != null ? _resources.doubles[name] : null);
  if (value is String && value.startsWith(prefix)) {
    // Refer to other resource
    String referName = value.substring(prefix.length);
    soughtNames ??= Set<String>();
    soughtNames.add(name);
    if (!soughtNames.contains(referName)) {
      soughtNames.add(referName);
      return getDouble(
        referName,
        fromResources: fromResources,
        soughtNames: soughtNames,
      );
    } else if (_notBaseResources(fromResources)) {
      // The name has been sought in current resource file,
      // If the current resource is not the base resource,
      // seek the base resource.
      soughtNames.clear();
      return getDouble(
        referName,
        soughtNames: soughtNames,
      );
    }
  }
  if (value is num) {
    return value.toDouble();
  }
  return 0;
}

bool _notBaseResources(Resources fromResources) =>
    fromResources != null && fromResources != _resources;


example/lib/main.dart
/*
 * Copyright 2019 chrisfang6
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:resource_manager/resource_manager.dart';
import 'package:resource_manager/resources.dart';
import 'style.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Example for Resource Manager'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Resources _exampleResources;

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _fetchResources();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: _buildContent(),
    );
  }

  Future _fetchResources() async {
    await initBaseResources(window.devicePixelRatio, 'res/values.json');
    _exampleResources ??= await getResources('res/example.json');
    // Update screen
    setState(() {});
  }

  Widget _buildContent() {
    if (_exampleResources == null) {
      return Center(
        child: Text(
          'Loading...',
          style: TextStyle(
            color: Color(0xffff0000),
            fontSize: 50,
            fontWeight: FontWeight.bold,
          ),
        ),
      );
    } else {
      return ListView.builder(
        itemCount: 23,
        padding: EdgeInsets.all(
          getDimension("list_padding", fromResources: _exampleResources),
        ),
        itemBuilder: (context, position) {
          return _getItem(position);
        },
      );
    }
  }

  Widget _getItem(int position) {
    var fontSize = "H1";
    var fontWeight = "Regular";
    var fontColor = "Royal Blue";
    switch (position) {
      case 0:
        return _getColorTable();
      case 1:
        fontWeight = "Bold";
        return Padding(
          padding: EdgeInsets.only(
              top: getDimension("gap", fromResources: _exampleResources)),
          child: _getText(
              fontSize, fontWeight, fontColor, "H1 color", getH1BoldTextStyle),
        );
      case 2:
        fontWeight = "Medium";
        return _getText(
            fontSize, fontWeight, fontColor, "H1 color", getH1MediumTextStyle);
      case 3:
        fontWeight = "Light";
        return _getText(
            fontSize, fontWeight, fontColor, "H1 color", getH1LightTextStyle);
      case 4:
        return _getText(
            fontSize, fontWeight, fontColor, "H1 color", getH1RegularTextStyle);
      case 5:
        fontSize = "H2";
        return _getText(
            fontSize, fontWeight, fontColor, "H2 color", getH2RegularTextStyle);
      case 6:
        fontSize = "H3";
        return _getText(
            fontSize, fontWeight, fontColor, "H3 color", getH3RegularTextStyle);
      case 7:
        fontSize = "H4";
        return _getText(
            fontSize, fontWeight, fontColor, "H4 color", getH4RegularTextStyle);
      case 8:
        fontSize = "H5";
        return _getText(
            fontSize, fontWeight, fontColor, "H5 color", getH5RegularTextStyle);
      case 9:
        fontSize = "H6";
        return _getText(
            fontSize, fontWeight, fontColor, "H6 color", getH6RegularTextStyle);
      case 10:
        return Padding(
          padding: EdgeInsets.only(
              top: getDimension("gap", fromResources: _exampleResources)),
          child: Text("Original image LOGO from base resource",
              style: getH2MediumTextStyle(Color(getColor("image label color",
                  fromResources: _exampleResources)))),
        );
      case 11:
        return Image.asset(getDrawable("logo"));
      case 12:
        return Text("Override image LOGO from base resource",
            style: getH2MediumTextStyle(Color(getColor("image label color",
                fromResources: _exampleResources))));
      case 13:
        return Image.asset(
            getDrawable("logo", fromResources: _exampleResources));
      case 14:
        return Padding(
          padding: EdgeInsets.only(
              top: getDimension("gap", fromResources: _exampleResources)),
          child: Text(
              "integer 1 from base file is " +
                  getInteger("integer 1").toString() +
                  ", and from example file is " +
                  getInteger("integer 1", fromResources: _exampleResources)
                      .toString() +
                  " as it refers to the one in base file",
              style: getH4MediumTextStyle(Color(getColor("num label color",
                  fromResources: _exampleResources)))),
        );
      case 15:
        return Text(
            "integer 2 from base file is " +
                getInteger("integer 2").toString() +
                ", and from example file is " +
                getInteger("integer 2", fromResources: _exampleResources)
                    .toString(),
            style: getH4MediumTextStyle(Color(getColor("num label color",
                fromResources: _exampleResources))));
      case 16:
        return Text(
            "double 1 from base file is " +
                getDouble("double 1").toString() +
                ", and from exmaple file is " +
                getDouble("double 1", fromResources: _exampleResources)
                    .toString() +
                " as it refers to the one in base file",
            style: getH4MediumTextStyle(Color(getColor("num label color",
                fromResources: _exampleResources))));
      case 17:
        return Text(
            "double 2 from base file is " +
                getDouble("double 2").toString() +
                ", and from example file is " +
                getDouble("double 2", fromResources: _exampleResources)
                    .toString(),
            style: getH4MediumTextStyle(Color(getColor("num label color",
                fromResources: _exampleResources))));
      case 18:
        return Padding(
          padding: EdgeInsets.only(
              top: getDimension("gap", fromResources: _exampleResources)),
          child: Text(
              "Device Pixel Ratio is " + window.devicePixelRatio.toString(),
              style: getH2MediumTextStyle(Color(getColor("pixel label color",
                  fromResources: _exampleResources)))),
        );
      case 19:
        var fsize = getDimension("shown dp", fromResources: _exampleResources);
        return Text("50dp: shown dp is " + fsize.toString(),
            style: TextStyle(
                fontSize: fsize,
                color: Color(getColor("pixel label color",
                    fromResources: _exampleResources))));
      case 20:
        var fsize = getDimension("shown px", fromResources: _exampleResources);
        return Text("50px: shown dp is " + fsize.toString(),
            style: TextStyle(
                fontSize: fsize,
                color: Color(getColor("pixel label color",
                    fromResources: _exampleResources))));
      case 21:
        return Padding(
          padding: EdgeInsets.only(
              top: getDimension("gap", fromResources: _exampleResources)),
          child: Text("String 1 from base file is " + getString("string 1"),
              style: getH2MediumTextStyle(Color(getColor("string label color",
                  fromResources: _exampleResources)))),
        );
      case 22:
        return Text(
            "String 1 from example file is " +
                getString("string 1", fromResources: _exampleResources),
            style: getH2MediumTextStyle(Color(getColor("string label color",
                fromResources: _exampleResources))));
      default:
        return Text("unknown widget");
    }
  }

  Text _getText(String fontSize, String fontWeight, String fontColor,
      String color, TextStyle Function(Color color) getTextStyle) {
    return Text(
      _getDisplayText(fontSize, fontWeight, fontColor),
      style: getTextStyle(
          Color(getColor(color, fromResources: _exampleResources))),
    );
  }

  Widget _getColorTable() => Table(
        columnWidths: const <int, TableColumnWidth>{
          0: IntrinsicColumnWidth(),
          1: FlexColumnWidth(1.0)
        },
        defaultVerticalAlignment: TableCellVerticalAlignment.middle,
        children: [
          _getColorTableRow("FireBrick"),
          _getColorTableRow("DeepPink"),
          _getColorTableRow("Gold"),
          _getColorTableRow("OliveDrab"),
          _getColorTableRow("Lime"),
          _getColorTableRow("RoyalBlue"),
          _getColorTableRow("DarkOrchid"),
          _getColorTableRow("As FireBrick"),
        ],
      );

  TableRow _getColorTableRow(String colorName) => TableRow(children: [
        Text(
          colorName,
          style: getH3BoldTextStyle(
            Color(getColor("black")),
          ),
        ),
        Padding(
          padding: EdgeInsets.all(
            getDimension("color_bar_padding", fromResources: _exampleResources),
          ),
          child: Container(
            height: getDimension("color_bar_height",
                fromResources: _exampleResources),
            color: Color(getColor(colorName, fromResources: _exampleResources)),
          ),
        )
      ]);

  String _getDisplayText(
          String fontSize, String fontWeight, String fontColor) =>
      "$fontSize $fontWeight " + "text with color" + " $fontColor";
}
